Add Feature Status documentation (#176)
argoproj/argoproj: community/feature-status.md

# Feature Status

> This document describes the different status a given feature can have
> in Argo projects.

Reviewer: Plural?
```suggestion
This document describes the different statuses a given feature can have
in Argo projects.
```

Author: fixed

## Overview

Argo umbrella projects are maintained by the open-source community.
This means that features are provided by different companies with
different use-cases. While certain features are battle tested by
corporations with large infrastructure, this is not always the case.
Feature status is a classification that can be used by Argo projects
to clarify and set expectations about a given functionality. However,
not every feature will have to be associated with a given status.
Maintainers will use this classification as a framework and gradually
associate features with the proper current status. When available, the
current status will be informed in the documentation of that specific
feature.

## Definition

Argo projects adopt the standard software life cycle terminology to
define the different feature statuses:

- Alpha
- Beta
- Stable

### Alpha

Alpha means that a feature is in its early stage and is considered as
experimental. It can be removed from future releases followed from a

> deprecation stage. This is a useful status to use when there are

Author: changed according to the suggestion

uncertainties about a new feature and maintainers want to get feedback
from the community. Alpha features are usually behind feature flags
disabled by default. Users are expected to understand the risks of

> enabling Alpha features in production environment.

Author: Agreed. Removed

### Beta

Beta features are usually enabled by default and are not considered as
experimental. Beta features are considered in its early maturity
stage. This means that common use cases are functional but there are
known uncovered edge cases. Edge cases will be tracked as issues and

> are expected to be voted. In most cases, highly voted issues will take

Reviewer: Would it also make sense to clarify how voting takes place? I assume it's the number of emoji reactions on the PR description.

> are expected to be voted. In most cases, highly voted issues will take

Author: Good idea. added the voting system clarification.

higher priority. Beta features may not be backwards compatible with
previous releases while in this status.

### Stable

> Stable features are tested and validated with a variety of use cases.
> Stable features are considered mature and are expected to be backwards
> compatible. Issues will be tracked as bugs and depending on the impact

Reviewer: Remove the redundant "stable features" in favour of commas?
```suggestion
Stable features are tested and validated with various use cases,
are considered mature, and are expected to be backwards compatible.
Issues will be tracked as bugs and depending on the impact
```

a patch with the fix might be released.

## The promotion process

> Features can be introduced in Argo projects in Alpha or Beta statuses.

Author: good idea. changed all occurrences

After a feature is introduced with a specific status it can then be
promoted to the next one.

The transition from one status to the other respect the following
rules:

1. Alpha features can transition to Beta.
1. If for some reason an Alpha feature fails to be promoted, it will
eventually be removed from the code base.

> 1. Beta features can only transition to Stable.

Reviewer 2: This makes it seem like beta features have to go stable I think that a beta feature still has potential to be removed it is just more unlikely. I don't think we should corner ourself into not being able to remove a beta feature it follows k8s appoarch just without the hard deadlines etc etc.

Reviewer 2: Yes I think that alpha is always behind feature flag and disabled by default. Yup and I think it makes sense for beta to be enabled by default but maintain the feature flag so that if there is a bug end users can just turn it off as a work around etc. I think we should have the requirement to keep the feature flag until GA just to provide users workaround in case of nasty bugs in beta etc.

Author: Agreed. This is something we can probably do for new beta features. But for all existing ones, it would be a huge effort to introduce FF for all of them. That's why I'm a bit hesitant to set too many things in stone in this doc and not leave room for some flexibility.

Author: updated

The diagram bellow represents the possible transitions features can
have:

```
                    ┌───────┐
  New Feature'─────►│ Alpha ├───► Removed
                    └───┬───┘
                        │
                        ▼
                     ┌──────┐
  New Feature"──────►│ Beta │
                     └──┬───┘
                        │
          ┌────────┐    │
          │ Stable │◄───┘
          └────────┘
```

Promoting a feature is a responsibility shared between maintainers,
contributors and the user community. Any of those parties can request
a feature to be promoted. Ideally this can be done by creating an
issue or a discussion in the Github repository of the project
containing that feature. Promotion requests can be brought to
maintainers attention in our weekly [Argo Contributors Office
hours][1] meeting.

## What it means for Argo users

Feature status was introduced as a framework to provide transparency

> and allow setting up expectations for Argo users. It should be used to

Author: In this case we are talking about one framework which is called Feature Status. Changed the text to make it "sound" better. ptal

support the decision making process about whether or not a given
feature is feasible to be leveraged by a company or individual. Argo
projects depend on the user base to provide valuable feedback and

> submit bugs with good information so features and be promoted to a

Reviewer: ```suggestion
submit bugs with good information so features can be promoted to a
```

> submit bugs with good information so features and be promoted to a

Author: Fixed. Tks!

more mature stage.

## What it means for Argo contributors

> 

Author: Agree. We can try that. It will require discipline from every approver to make sure a new feature is classified properly.

> 

Author: updated

If you are contributing with new feature in one of the Argo projects,

> please consider classifying it in one of the statuses described in

Author: Yes. The status is reviewed by an Argo Approver during the PR review process. Not free for all. :)

this document. The current status should be indicated next to every
feature in our official documentation.

The format of a feature status should be: `Current Status: <status>

> (Since <version>)`

Reviewer 3: Does it need clarifying that "Since" means the version the feature was introduced, not the version it changed status?

Author: makes sense. added. tks!

# Feature Status

This document describes the different statuses a given feature can
have in Argo projects.

## Overview

Argo umbrella projects are maintained by the open-source community.

> This means that features are provided by different companies with

Reviewer 4: "companies" --> "organizations" ?

different use-cases. While certain features are battle tested by
corporations with large infrastructure, this is not always the case.
Feature status is a classification that can be used by Argo projects
to clarify and set expectations about a given functionality. However,
not every feature will have to be associated with a given status.
Maintainers will use this classification as a framework and gradually
associate features with the proper current status. When available, the
current status will be informed in the documentation of that specific
feature.

## Definition

Argo projects adopt the standard software life cycle terminology to
define the different feature statuses:

- _Alpha_
- _Beta_

> - _Stable_

Reviewer 4: Wondering if we should call it "GA" instead of "Stable".

Author: Good point. I think I have a slight preference for `Stable` to avoid the acronym but I am not strong opinionated about that.

Reviewer: +1 for `Stable`. It's also the rhetoric the doc versioning uses.

Feature status is a framework to provide transparency and allow
setting expectations for Argo users. It should be used to support the
decision making process about whether or not a given feature is
feasible to be leveraged by a company or individual. Argo projects
depend on the user base to provide valuable feedback and submit bugs
with good information so features can be promoted to a more mature
stage.

## What it means for Argo contributors

If you are contributing with new feature in one of the Argo projects,
please classify it with one of the statuses described in this
document. The current status should be indicated next to every feature
in our official documentation. The status must also be informed in the
title of the PR introducing the feature so it can be available in the
release notes.

The format of a feature status in the PR title should be:

`feat: <title> (<status>)`

Example:

`feat: My awesome new feature description (Alpha)`

The format of a feature status in the documentation should be:

`Current Status: <status> (Since <version>)`

Since relates to the version a feature was introduced, not the version
it changed the status.

Example:

> `## Server-Side Apply Current Status: Beta (Since v2.5.0)`

Reviewer 4: When an alpha feature is getting promoted into beta, do we expect a contributor to make a PR titled "feat: my previously built cool feature (beta)" ?

Author: Not necessarily. I needs to be updated is the page where the feature is documented to point to the current status.

Author: @sbose78 On a second thought, a PR will have to be created anyways to update at least the documentation. I think we should have the PR title following the pre-defined format in this case as well. Updated the docs. 
PTAL

Reviewer 4: Thank you @leoluz , looks great.

associate features with the proper current status. When available, the
current status will be informed in the documentation of that specific
feature.

## Definition

Argo projects adopt the standard software life cycle terminology to
define the different feature statuses:

- _Alpha_
- _Beta_
- _Stable_

### Alpha

_Alpha_ means that a feature is in its early stage and is considered
as experimental which means that it can be removed from future
releases. This is an useful status to use when there are uncertainties
about a new feature and maintainers want to get feedback from the
community. _Alpha_ features are usually behind feature flags disabled
by default. Users are expected to understand the risks of enabling
_Alpha_ features.

### Beta

_Beta_ features are usually enabled by default and are not considered
as experimental. _Beta_ features are considered in its early maturity
stage. This means that common use cases are functional but there are
known uncovered edge cases. Edge cases will be tracked as issues and
are expected to be voted on. Votes take place by adding a thumbs up
emoji reaction in the main description of an issue. In most cases,

> issues with more votes will take higher priority. _Beta_ features may
> not be backwards compatible with previous releases while in this
> status.

Reviewer 5: To make clear that beta features aren't allowed to break backwards-compatibilty _in general_, just within the feature itself.

```suggestion
issues with more votes will take higher priority. _Beta_ features may
not be backwards compatible with previous releases of the feature
while in this status.
```

Author: changed. Tks!

### Stable

_Stable_ features are tested and validated with various use cases, are
considered mature, and are expected to be backwards compatible. Issues
will be tracked as bugs and depending on the impact a patch with the
fix might be released.

## The promotion process

> Features can be introduced in Argo projects in _Alpha_ or _Beta_
> statuses. After a feature is introduced with a specific status it can
> then be promoted to the next one.

Reviewer 5: ```suggestion
Features can be introduced in Argo projects in _Alpha_ or _Beta_
statuses. If a feature is not introduced with a particular status,
it is assumed to be stable. After a feature is introduced with a 
specific status it can then be promoted to the next one.
```

Author: updated tks

a feature to be promoted. Ideally this can be done by creating an
issue or a discussion in the Github repository of the project
containing that feature. Promotion requests can be brought to
maintainers attention in our weekly [Argo Contributors Office
hours][1] meeting.

## What it means for Argo users

Feature status is a framework to provide transparency and allow
setting expectations for Argo users. It should be used to support the
decision making process about whether or not a given feature is
feasible to be leveraged by a company or individual. Argo projects
depend on the user base to provide valuable feedback and submit bugs
with good information so features can be promoted to a more mature
stage.

## What it means for Argo contributors

If you are contributing with new feature in one of the Argo projects,
please classify it with one of the statuses described in this
document. The current status should be indicated next to every feature
in our official documentation. The format of a feature status in the
documentation should be:

`Current Status: <status> (Since <version>)`

Since relates to the version a feature was introduced, not the version

> it changed the status.

Reviewer 5: I think both things are useful, but "when it became stable" is more useful.
